src/icrar/leap-accelerate/ms/MeasurementSet.cc

```diff
@@ -32,30 +34,102 @@ namespace icrar
     , m_filepath(filepath)
     , m_readAutocorrelations(readAutocorrelations)
     {
-        m_stations = overrideNStations.is_initialized() ? overrideNStations.get() : m_measurementSet->antenna().nrow();
+        // Check and use unique antennas 
+        casacore::Vector<casacore::Int> a1 = m_msmc->antenna1().getColumn();
+        casacore::Vector<casacore::Int> a2 = m_msmc->antenna1().getColumn();
+        auto a1s = std::set<int32_t>(a1.cbegin(), a1.cend());
+        auto a2s = std::set<int32_t>(a2.cbegin(), a2.cend());
+        std::set_union(a1.cbegin(), a1.cend(), a2.cbegin(), a2.cend(), std::inserter(m_antennas, m_antennas.begin()));
+        
+        if(overrideNStations.is_initialized())
+        {
+            m_stations = overrideNStations.get();
+        }
+        else if(m_antennas.size() != m_measurementSet->antenna().nrow())
+        {
+            LOG(warning) << "ms antennas = " << m_measurementSet->antenna().nrow();
+            LOG(warning) << "unique antennas = " << m_antennas.size();
+            LOG(warning) << "using unique antennas";
+            m_stations = m_antennas.size();
+        }
+        else
+        {
+            m_stations = m_measurementSet->antenna().nrow();
+        }
+
+        Validate();
     }
 
-    MeasurementSet::MeasurementSet(const casacore::MeasurementSet& ms, boost::optional<int> overrideNStations, bool readAutocorrelations)
-    : m_measurementSet(std::make_unique<casacore::MeasurementSet>(ms))
-    , m_msmc(std::make_unique<casacore::MSMainColumns>(*m_measurementSet))
-    , m_msc(std::make_unique<casacore::MSColumns>(*m_measurementSet))
-    , m_readAutocorrelations(readAutocorrelations)
+    void MeasurementSet::Validate() const
     {
-        m_stations = overrideNStations.is_initialized() ? overrideNStations.get() : m_measurementSet->antenna().nrow();
-    }
+        //Stations
+        casacore::Vector<casacore::Int> a1 = m_msmc->antenna1().getColumn();
+        casacore::Vector<casacore::Int> a2 = m_msmc->antenna1().getColumn();
+        auto a1s = std::set<int32_t>(a1.cbegin(), a1.cend());
+        auto a2s = std::set<int32_t>(a2.cbegin(), a2.cend());
+        std::set<std::int32_t> antennas;
+        std::set_union(a1.cbegin(), a1.cend(), a2.cbegin(), a2.cend(), std::inserter(antennas, antennas.begin())); 
+        
+        if(antennas.size() != GetNumStations())
+        {
+            LOG(error) << "unique antennas does not match number of stations";
+            LOG(error) << "unique antennas: " << antennas.size();
+            LOG(error) << "stations: " << GetNumStations();
+            //throw exception("number of stations incorrect", __FILE__, __LINE__);
+        }
```

Comment:
Looking at this makes me think that the number of stations declared in the MS are the ground truth, and that the `-s` command line option seems redundant. Is that reading correct?

Reply:
I think this is meant to be a warning message. I added this one as an attempt to help hint whether the measurement should be treated as `--stations 102 --autocorrelations false` rather than `--stations 101 --autocorrelations true`, both cases end up calculating the same result currently. I'll take another look confirm whether this message helps.